cosmos-reflection: extend to support writing by reflection clients

fdymylja opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review.
-->

closes: https://github.com/cosmos/cosmos-sdk/issues/8959

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

fdymylja commented:
@aaronc, @amaurym this would be the proposed addition to support writing reflection clients

fdymylja commented:
pinging also @robert-zaremba in case you wanna provide some feedback and insights on the proposed API

fdymylja commented:
introduces a breaking change in proto because golang package has changed, to make it consistent to where gRPC reflection lives. 

on a second note, I had tried a lot to give even stronger reflection to the point in which client would not need to rely on auth for building txs. unfortunately current SDK abstractions and structure of signing (partly in types, partly in auth) does not allow for stronger reflection capabilities. they could have been achieved but the approach would have not been as clean as I would have liked. Hence post-poning this when we do more refactoring work on the sdk.

This PR can be merged and this would allow clients to write on any chain, just by relying on the tx.Tx type and auth.

alessio reviewed (approved):
Great work. ACK from me.

jgimeno reviewed (approved):
Top!

fdymylja commented:
@amaurym, @robert-zaremba @aaronc sorry to ping you again, the PR is ready for review, if you have any feedback it'd be greatly appreciated

amaurym commented:
Sorry for gettinig back so late. I'll have a look at this tomorrow morning, thanks for the PR!

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+  string id = 1;
+}
+
+// CodecDescriptor describes the registered interfaces and provides metadata information on the types
+message CodecDescriptor {
+  // interfaces is a list of the registerted interfaces descriptors
+  repeated InterfaceDescriptor interfaces = 1;
+}
+
+// InterfaceDescriptor describes the implementation of an interface
+message InterfaceDescriptor {
+  // fullname is the name of the interface
+  string fullname = 1;
+  // interface_accepting_messages contains information regarding the proto messages which contain the interface as google.protobuf.Any field
+  repeated InterfaceAcceptingMessageDescriptor interface_accepting_messages = 2;
+  // interface_implementers is a list of the descriptors of the interface implementers
+  repeated InterfaceImplementerDescriptor interface_implementers = 3;
+}
+
+// InterfaceImplementerDescriptor describes an interface implementer
+message InterfaceImplementerDescriptor {
+  // fullname is the protobuf queryable name of the interface implementer
+  string fullname = 1;
+  // type_url defines the type URL used when marshalling the type as any
+  // this is required so we can provide type safe google.protobuf.Any marshalling and
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 2;
+}
+
+// InterfaceAcceptingMessageDescriptor describes a protobuf message which contains
+// an interface represented as a google.protobuf.Any
+message InterfaceAcceptingMessageDescriptor {
+  // fullname is the protobuf fullname of the type containing the interface
+  string fullname = 1;
+  // field_descriptor_names is a list of the protobuf name (not fullname) of the field
+  // which contains the interface as google.protobuf.Any (the interface is the same, but
+  // it can be in multiple fields of the same proto message)
+  repeated string field_descriptor_names = 2;
+}
```
Isn't this redundant with the existing `ListAllInterfaces` and `ListImplementations` methods?

fdymylja replied:
This adds a lot more safety when it comes to building marshalers and unmarshalers for types which hold any types. So we can make sure that only certain types can be resolved on certain proto.Messages. For example you don't want a proto.Message of a sdk.ServiceMsg to be resolved in a field which is supposed to accept interfaces.

aaronc replied:
But all of this information is in the proto descriptors themselves so why duplicate this? Clients could get the proto descriptors from grpc reflection and pull out these annotations. I guess you're trying to simplify that step?

fdymylja replied:
yes, it's meant to be a semplification which wouldnt need clients to rely on specific versions of the cosmos proto which expose this. On a second note, the annotation does not contain information on the protobuf name of the interface, example: annotation is `AccountI`, codec resolvable name is `cosmos.auth.v1beta1.AccountI`, unluckily due to this, the object here exists but cannot yet be used. :(

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -3,10 +3,14 @@ package cosmos.base.reflection.v1beta1;
+// ReflectionService defines a service for application reflection.
 service ReflectionService {
+  // GetAppDescriptor returns the full cosmos application descriptor
+  rpc GetAppDescriptor(GetAppDescriptorRequest) returns (GetAppDescriptorResponse) {
```
Rather than a single `AppDescriptor` I would personally prefer individual methods for each type of information. Not a strongly held preference, but to me that would feel cleaner.

fdymylja replied:
I thought about this too, but when experimenting in the end you'd need to get sequentially all this descriptor to be able to  write on a chain, but we can split them.

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+
+// AuthnDescriptor provides information on how to sign transactions without relying
+// on the online RPCs GetTxMetadata and CombineUnsignedTxAndSignatures
+message AuthnDescriptor {
```
Why not `SignModeDescriptor` or `AuthenticationDescriptor`

fdymylja replied:
I thought it fitted better the intention of renaming auth to authn

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+
+// ServiceMsgDescriptor describes an sdk.ServiceMsg type
+message ServiceMsgDescriptor {
+  // fullname is the protobuf fullname of the given sdk.ServiceMsg
+  string fullname = 1;
+  // route is the sdk.ServiceMsg route, it is equal to type_url
+  string route = 2;
+  // type_url defines the type URL used when marshalling the type as any
+  // this is required so we can provide type safe google.protobuf.Any marshalling and
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 3;
```
These are all redundant as far as I know. We just need the type URL.

fdymylja replied:
The concrete name is required too for marshal options to get the correct message descriptor which is saved via protobuf.Fullname, type URL is custom to our codec implementation

aaronc replied:
Can you give me an example of what all three of these would be for a Msg?

fdymylja replied:
Yes sure. For example when I resolve the ServiceMsg interface. I get the following type URL: `/cosmos.crisis.v1beta1.Msg/VerifyInvariant` 

Now if I am a reflection client, how do I know which is the protobuf fullname for this typeURL? The protobuf fullname is actually: `cosmos.crisis.v1beta1.MsgVerifyInvariant` so I can now query the gRPC reflection to give me the file which contains this symbol. 

Here you have an example of the app descriptor for simapp: https://pastebin.com/BxSyKXJk

aaronc replied:
Well there's a distinction in ServiceMsg's between the method and the request type. What you're calling the full name is actually the request type name. I think we should have two fields for request type full name and response type full name for service msgs.

aaronc replied:
Also ServiceMsg's do abuse a bit the type URL of Any, that's covered in ADR 031 as sort of a necessary "abuse" so this can happen. Hope it's clear enough what's actually happening.

fdymylja replied:
Sorry I didn't understand, so do you agree that we need both typeURL and protobuf fullname (the name the type is registered with in the protobuf registry, not our codec) so when we marshal the serviceMsg as anypb it will have the correct type URL, and the protobuf fullname so that we can build the correct message via descriptors?

> I think we should have two fields for request type full name and response type full name for service msgs.

I don't know how to fetch those, currently I don't think anywhere in the app we save the `Msg` server service descriptors?

fdymylja replied:
Side question: why in ADR-031 we preferred the custom type URL, instead of the `MsgServer` service descriptor method fullname?

fdymylja replied:
the configurator which holds `MsgServer` descriptors is defined at specific app level (not `baseapp` level), so the reflection server cannot make use of that. I've defined in the `ServiceMsgDescriptor` the `response_fullname` but as of now we cannot make use of that. After the configurator finds its place in `baseapp` we can fill the missing information.

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+}
+
+// LegacyMsgDescriptor describes an sdk.Msg type
+message LegacyMsgDescriptor {
+  // fullname is the protobuf fullname of the message
+  string fullname = 1;
+  // type_url defines the type URL used when marshalling the type as any
+  // this is required so we can provide type safe google.protobuf.Any marshalling and
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 2;
+}
```
The plan that was announced with Stargate is that we would be deprecating legacy messages. I would rather not backtrack on this and I'd rather not additional support for things that are going to be deprecated.

fdymylja replied:
Ok, but are all supported msgs already service msgs? Because I'd like reflection clients to be able to handle anything that a chain can accept via handlers.

aaronc replied:
All supported Msg's can be represented as service msgs now

fdymylja replied:
If all supported msgs are already servicemsgs then this does not need to exist, I thought ATM we were supporting both in a separate way. Thanks for the heads up.

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+// TxDescriptor describes the accepted transaction type
+message TxDescriptor {
+  // fullname is the protobuf fullname of the raw transaction type (for instance the tx.Tx type)
+  string fullname = 1;
```
Maybe this is okay but I kind of have mixed feelings about explicitly supporting non-standard transaction types. As we were going through the Stargate migration, it was my feeling (and I believe other's as well) that making `sdk.Tx` an interface and supporting multiple transaction types in the ante handler layer is something that adds a lot of overhead without adding a lot of value. I'm not saying I'm totally opposed to keeping things as is and allowing multiple Tx types and describing that with reflection here BUT it does feel counter to a lot of the learnings from Stargate. In particular, if chains support different Tx types all sorts of things like block explores and wallets become incompatible between Cosmos chains. I'd personally rather see us standardize on one `Tx` type and not support polymorphism here.

fdymylja replied:
I know, this is not intended to be used to support polymorphysm, it's meant to signal to reflection clients if they can write on the chain. 
Ex: if we make a breaking change and decide to switch the tx.Tx object, then a reflection client would know it cannot write to this chain because it can support `cosmos.base.v1beta1.Tx` but not `cosmos.base.v2beta1.Tx`

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+message SigningModeDescriptor {
+  // name defines the unique name of the signing mode
+  string name = 1;
+  // authn_info_provider_method_fullname defines the fullname of the method to call to get
+  // the metadata required to authenticate using the provided sign_modes
+  string authn_info_provider_method_fullname = 2;
```
What is this for?

fdymylja replied:
this is a left over of my attempts to not force reflection clients to rely on auth.

Example: I wanna get the metadata required to sign with sign_mode_direct, which are acc num, sequence and chainID then I can provide to the reflection client an endpoint to query for this info.

currently auth does not support this, although I feel it should. Or generally speaking auth should be abstracted better in the sdk.

aaronc replied:
In general there should be a simple service to get acc num and sequence independent of how the auth module represents accounts. As you probably know, I think `x/auth` is way too complicated. But in the short term we should have a simple generic "get auth info" service somewhere...

fdymylja replied:
Yes, I wanted to implemented this specific endpoint, but I didn't know where to put it to be honest, should it be in auth? I don't think so. I think it should be in base and auth should implement it. But then it would be inconsistent with how we deal with authentication which is partly in base and partly in auth. I think this deserves an aside conversation which fits your discussion: https://github.com/cosmos/cosmos-sdk/discussions/9011.

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+// to sign a message given a SigningModeDescriptor, but it's better to think about
+// this another time
+message SigningModeDescriptor {
+  // name defines the unique name of the signing mode
+  string name = 1;
```
Can we instead reuse the `SIGN_MODE` enum directly here?

fdymylja replied:
Do we want to import into a reflection server a dependency on base?

aaronc replied:
So I think there's a scoping issue here. I would say this should actually be in `cosmos.tx.v1.beta1.Service`. Maybe it's enough to describe what the tx service URL is and then we can get the sign modes from the tx service so you don't need to import.

fdymylja replied:
I do agree with what you're saying. But at the same time this would need the reflection client to rely not only on reflection and auth but on the tx service too. Uhm... I don't know

amaurym replied:
I always understood "reflection" as "a proto service returning something about the proto files themselves". So strictly speaking, every field in this file should be a proto fullname string. With this mindset, I would also put this info in the TxService.

For example, here the TxDescriptor would only return proto fullnames:
- `fullname string` -> return `cosmos.tx.v1beta1.Tx`
- `service_name string` -> return `cosmos.tx.v1beta1.Service`
- `repeated MsgDescriptor msgs` -> return all the Msg Services fullnames

and other normal services can return whatever they want.

It seems that this PR understands "reflection" as "proto reflection + any chain config that's not queryable from state". Which is fine, but yeah, creates scoping issues as whether `chain-id`/`codecs`/`sign_modes`/`bech32` belong to normal services or reflection services.

aaronc reviewed (dismissed):
I really like the motivation here @fdymylja, but I feel like this API could be greatly simplified and likely we don't need to support all of what is here. I left many comments inline, but at a high level I'll summarize what I think should be added to the reflection service:
* chain ID
* bech32 account prefix
* available sign modes
* Tx service methods

I believe most of the other stuff is already supported elsewhere (i.e. the existing `ListAllInterfaces`, etc. methods cover the codec use case and grpc reflection covers queries), or is not really something we want to add support for (i.e. legacy msg's and queriers, legacy global config settings, etc.).

fdymylja commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+}
+
+// QueryServiceDescriptor describes a cosmos-sdk queryable service
+message QueryServiceDescriptor {
```
This covers only the gRPC queriers, it only adds one more thing: the tendermint query path, which can be used if a client desires to query via tendermint rather than gRPC. Plus it holds a space in case in the future we desire to add more information on specific query services. For example: module name would be useful to identify to which module that query method belongs to, without needing to parse the servicedescriptor fullname

amaurym replied:
> it only adds one more thing: the tendermint query path, which can be used if a client desires to query via tendermint rather than gRPC.

But the tendermint query path (`full_query_path` below) is exactly the gRPC method fullname.

I would remove `repeated QueryMethodDescriptor methods = 2`, we just need `string fullname = 1` and you can further reflection to list all methods.

fdymylja commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+  string bech32_validator_public_key_prefix = 5;
+  // bech32_consensus_public_key_prefix is the bech32 prefix for consensus nodes public keys
+  string bech32_consensus_public_key_prefix = 6;
+  // purpose is the BIP-0044 purpose code
+  uint32 purpose = 7;
+  // coin_type is the BIP-0044 coin type code
+  uint32 coin_type = 8;
+  // full_fundraiser_path is the BIP-0044 prefix (deprecated in favor of full_bip44_path)
+  string full_fundraiser_path = 9;
+  // full_bip44_path is the BIP-0044 prefix
+  string full_bip44_path = 10;
```
Ok, I'll get rid of it.

fdymylja commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+  string bech32_validator_address_prefix = 2;
+  // bech32_consensus_address_prefix is the bech32 prefix for consensus nodes addresses
+  string bech32_consensus_address_prefix = 3;
+  // bech32_account_public_key_prefix is the bech32 prefix for accounts public keys
+  string bech32_account_public_key_prefix = 4;
+  // bech32_validator_public_key_prefix is the bech32 prefix for validators public keys
+  string bech32_validator_public_key_prefix = 5;
+  // bech32_consensus_public_key_prefix is the bech32 prefix for consensus nodes public keys
+  string bech32_consensus_public_key_prefix = 6;
```
Should those be marked as deprecated in your opinion, or should we get rid of them totally?

aaronc replied:
They've been removed in master afaik. @robert-zaremba ?

fdymylja replied:
I have the latest version of master on my branch but those methods are still there, maybe sdk.Config should be cleaned up?

fdymylja replied:
I've cleaned up the configuration descriptor, only bech32 acc addr prefix would be left

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -21,6 +25,187 @@ service ReflectionService {
+message ConfigurationDescriptor {
+  // bech32_account_address_prefix is the account address prefix
+  string bech32_account_address_prefix = 1;
+  // bech32_validator_address_prefix is the bech32 prefix for validator addresses
+  string bech32_validator_address_prefix = 2;
+  // bech32_consensus_address_prefix is the bech32 prefix for consensus nodes addresses
+  string bech32_consensus_address_prefix = 3;
```
We should add a query method to staking to get these prefixes. That is where these are used and required in msgs.

fdymylja replied:
If that's something specific to staking then the reflection client does not need that. I will get rid of it.

fdymylja commented:
@aaronc I've addressed your feedbacks. The only piece which I'm kind of not happy with is the `AuthnDescriptor`, but I think its development is dependent on other pieces of the sdk that I think this PR should not tackle. We could mark it as unstable or experimental. Let me know.

aaronc commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -3,10 +3,161 @@ package cosmos.base.reflection.v1beta1;
+
+// InterfaceImplementerDescriptor describes an interface implementer
+message InterfaceImplementerDescriptor {
+  // fullname is the protobuf queryable name of the interface implementer
+  string fullname = 1;
+  // type_url defines the type URL used when marshalling the type as any
+  // this is required so we can provide type safe google.protobuf.Any marshalling and
+  // unmarshalling, making sure that we don't accept just 'any' type
+  // in our interface fields
+  string type_url = 2;
```
I'm confused in general by your usage of both `fullname` and `type_url` here and in other places. From my perspective, there is only "type URL" OR "full name" in the protobuf world and we should use that consistently. There should not be cases where we have both a type URL and full name that could be different but refer to the same thing.

fdymylja replied:
@aaronc we do need both unfortunately.

When we build a protoregistry from gRPC reflection, the file descriptor will define, for example, the fullname of MsgSend as `cosmos.bank.v1beta1.MsgSend`, now if I want to resolve that type if it's wrapped as `google.protobuf.Any` I need to feed my protov2 type registry a way to give me my `proto.MessageType` from the given type URL which is going to be `/cosmos.bank.v1beta1.MsgSend`. 

Note: when I register a dynamic type from a file descriptor in the type registry, I cannot use a custom name. It's just going to take my MessageDescriptor from my FileDescriptor which defines its name. I don't want to edit raw file descriptors to make the (custom) any type URL match with the fullname.

aaronc replied:
This is not quite right I believe. As I said before `/cosmos.bank.v1beta1.Msg/Send` is the type URL for the method. `cosmos.bank.v1beta1.MsgSend` is the type URL for the _request type_. So what you want to register in the Any resolver is the request type of the service method. It is _incorrect_ to say that the " fullname of MsgSend as `cosmos.bank.v1beta1.MsgSend`".

fdymylja replied:
Then I'm not understanding sorry, as per ADR-031, I need to encode my MsgBankSend as:
`any{typeURL: /cosmos.bank.v1beta1.Msg/Send, value: proto.Marshal(MsgBankSend)}`

fdymylja replied:
This is the quote from ADR-031: 
```
In order to encode service methods in transactions, we encode them as Anys in the same TxBody.messages field as other Msgs. We simply set Any.type_url to the full-qualified method name (ex. /cosmos.gov.Msg/SubmitProposal) and set Any.value to the protobuf encoding of the request message (MsgSubmitProposal in this case).
``

aaronc replied:
Yeah, you encode and decode the request type of the method which happens to be `cosmos.v1beta1.bank.MsgSend`. That's the convention of ADR 031. We write `rpc(RequestType) returns (ResponseType)` for service methods. So you just need:
* method type URL
* request type URL
* response type URL

And you encode `Any(TypeURL: methodTypeUrl, Value: proto.Marshal(requestType))`

Does that help?

aaronc commented:
> @aaronc I've addressed your feedbacks. The only piece which I'm kind of not happy with is the `AuthnDescriptor`, but I think its development is dependent on other pieces of the sdk that I think this PR should not tackle. We could mark it as unstable or experimental. Let me know.

My general preference is to come up with APIs that we more or less feel good about before we commit to them... If we want to mark things as experimental and unstable, I think the right way would be a `v1alpha1` package and I would be totally fine having that alongside `v1` and `v1beta1` packages for the same larger package (i.e. `reflection`). So basically my preference is either to a) spend a bit more time working out the various kinks in this API design or b) push it into alpha if that's will be too slow of a process.

amaurym commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -3,10 +3,161 @@ package cosmos.base.reflection.v1beta1;
+
+// MsgDescriptor describes a cosmos-sdk message that can be delivered with a transaction
+message MsgDescriptor {
+  // msg contains a descriptor of sdk.ServiceMsg, note: sdk.Msg is not supported
+  // as every sdk.Msg is already an sdk.ServiceMsg. It is defined as a oneof in case
+  // different representation of a msg will be implemented.
+  oneof msg {
+    // service_msg is used when the message is an sdk.ServiceMsg type
+    ServiceMsgDescriptor service_msg = 1;
+  }
```
Not sure about the oneof, maybe we should just use ServiceMsgDescriptor and drop legacy Msgs altogether?

amaurym commented on proto/cosmos/base/reflection/v1beta1/reflection.proto:
```diff
@@ -3,10 +3,161 @@ package cosmos.base.reflection.v1beta1;
+}
+
+// ServiceMsgDescriptor describes an sdk.ServiceMsg type
+message ServiceMsgDescriptor {
+  // request_fullname is the protobuf fullname of the given sdk.ServiceMsg request
+  // this is the protobuf message type which should be used as google.protobuf.Any.value
+  // when delivering the msg to the DeliverTx endpoint
+  string request_fullname = 1;
+  // request_route is the sdk.ServiceMsg route, it is equal to type_url
+  string request_route = 2;
+  // request_type_url is the identifier that should be used as google.protobuf.Any.type_url
+  // when delivering the msg to the DeliverTx endpoint
+  string request_type_url = 3;
+  // response_fullname is the protobuf fullname of the given sdk.ServiceMsg response
+  string response_fullname = 4;
+}
```
IMO you just need one field here: string fullname of the service method. Then, you can use a separate reflection (e.g. the gogoreflection one) to query the request_fullname and the response_fullname of each method

fdymylja replied:
The configurator is not inside the BaseApp, hence we have no access to MsgServer descriptors. This is documented in the implementation.

amaurym reviewed (commented):
I kind of share the same feeling as @aaronc. This looks like a good start, but contains some confusing parts, which doesn't make an ideal user-facing API imo. To unblock immediate work on rosetta, we can merge this as v1alpha1.

fdymylja commented:
@amaurym, @aaronc, after further investigation the following are my conclusions:

TypeURL (we can discuss if we want to keep the field name as type_url or call it any_type_url, to further clarify its role) is a must that must be returned when we're querying for interfaces implementations:

- Sending Tx is not the only exception in which we represent Any's in a way that breaks Any spec.
- GetTx is another endpoint which requires the reflection clients.
- Any RPC (gRPC, or gRPC-gateway based) which accepts sdk.ServiceMsg as any forces us to use a custom type resolver to get the correct MessageDescriptor from that typeURL. 

How is any compatibility broken? Any spec states the following: `The last segment of the URL's path must represent the fully qualified name of the type (as in path/google.protobuf.Duration)`.

Now this aside making type resolving for any hard (which is why need to rely on typeURL field in the API proposed here), it breaks the functionalities of any itself. 
For example: in golang's Any implementation, we have the following: 
![image](https://user-images.githubusercontent.com/33157909/113414439-4d4e2100-93bd-11eb-90c5-80092e2a9ac7.png)

This is always gonna return false for sdk.ServiceMsgs even if the type we're trying to compare it to is correct.


This spec is broken for each proto.Message in the sdk that returns sdk.ServiceMsg interface wrapped as any. I suggest to revisit this part in ADR-031, and maybe opt for a custom type url which includes both the method and the correct fullname of the message.

As for: 
> It seems that this PR understands "reflection" as "proto reflection + any chain config that's not queryable from state". Which is fine, but yeah, creates scoping issues as whether chain-id/codecs/sign_modes/bech32 belong to normal services or reflection services.

I do agree, but it's too much of a struggle to propose 4 different services or changes in current APIs and then coordinate. Just agreeing on the API would block work which depends on this for months.

fdymylja commented:
I'm fine with reverting current changes to the reflection service and proposing this as a v2alpha1. WDYT?

amaurym commented:
> I'm fine with reverting current changes to the reflection service and proposing this as a v2alpha1. WDYT?

OK 👍

fdymylja commented:
@aaronc @amaurym the APIs are now split, the v1beta1 is reverted to how it was, v2alpha1 is aside and marked as experimental in the changelog too.

amaurym commented on server/grpc/server_test.go:
```diff
@@ -16,6 +16,8 @@ import (
 	"google.golang.org/grpc/metadata"
 	rpb "google.golang.org/grpc/reflection/grpc_reflection_v1alpha"
 
+	reflectionv1 "github.com/cosmos/cosmos-sdk/client/grpc/reflection"
```
I believe reflection goes belongs indeed more into `server`, not sure why it was initially in client. Should we move that folder to `server`?

fdymylja replied:
I can move it into server

amaurym commented on server/grpc/reflection/v2alpha1/reflection.go:
```diff
@@ -0,0 +1,219 @@
+		return nil, fmt.Errorf("unable to get *tx.Tx protobuf name")
+	}
+	// get msgs
+	msgImplementers := ir.ListImplementations(sdk.MsgInterfaceProtoName)
```
One idea I just thought (for the future) is creating new Protobuf options used for reflection, e.g.

```proto
// cosmos.bank.v1beta1
service Msg {
  option cosmos_msg_service = true;
  // all methods
}

service Query {
  option cosmos_query_service = true;
  // all methods
}
```

Which would avoid using `ListImplementations` for getting all ServiceMsgs. Just an idea.

fdymylja replied:
Yes, having this would sure help. Although I feel the best would be that the configurator becomes part of *BaseApp and we explicitly mark query services and msg services for the application.

amaurym reviewed (approved):
ACK from me

fdymylja commented on Makefile:
```diff
@@ -373,7 +373,7 @@ proto-all: proto-format proto-lint proto-gen
 proto-gen:
 	@echo "Generating Protobuf files"
-	$(DOCKER) run --rm -v $(CURDIR):/workspace --workdir /workspace tendermintdev/sdk-proto-gen:v0.1 sh ./scripts/protocgen.sh
+	docker run --rm -v $(CURDIR):/workspace --workdir /workspace tendermintdev/sdk-proto-gen:v0.1 sh ./scripts/protocgen.sh
```
Whoops! I was on windows yesterday and without this edit I couldn't build protos, then I must have committed this by mistake!

alessio commented:
Thanks so much @sahith-narahari! I'll merge as soon as I can

ytbigvinny288 reviewed (commented):
.

ytbigvinny288 reviewed (commented):
.

ytbigvinny288 reviewed (commented):
y

ytbigvinny288 reviewed (commented):
y

ytbigvinny288 reviewed (commented):
y

ytbigvinny288 reviewed (commented):
y
